proposals/lip-0043.md

```diff
@@ -330,39 +382,63 @@ mainchainRegParams = {
-      * `weights = [ validator.bftWeight for validator in validators ]` sorted according `keyList`.
-    * The `tag` is equal to `TAG_CHAIN_REG_MESSAGE`.
-    * The `netID` byte array corresponds to the network ID of the chain.
-    * The `aggregationBits` argument is the byte array given in `trs.params.aggregationBits`.
-    * The `signature` argument is the aggregate signature given in `trs.params.signature`.
-    * The `certificateThreshold` argument is equal to `getBFTParameters(b.header.height).certificateThreshold` where `b` is the current block.
-    * The `message` argument is the output bytes of the serialization, as specified in [LIP 0027][LIP-0027], of `trs.params.ownChainID`, `trs.params.ownName` and `trs.params.mainchainValidators` properties according to the following schema:
-
-    ```java
-    registrationSignatureMessage = {  
+```python
+def verify(trs: Transaction) -> None:
+    # The ownChainID property has to match with the chain identifier.
+    chainID = chain identifier of the current chain
+    if trs.params.ownChainID != chainID:
+        raise Exception("Invalid ownChainID property.")
+
+    # The ownName property has to contain only characters from the set [a-z0-9!@$&_.].
+    if not re.match(r"^[a-z0-9!@$&_.]+$", trs.params.ownName):
+        raise Exception("Invalid ownName property. It should contain only characters from the set [a-z0-9!@$&_.].")
+
+    # mainchainValidators must have exactly MAINCHAIN_NUMBER_ACTIVE_VALIDATORS elements.
+    if len(trs.params.mainchainValidators) != MAINCHAIN_NUMBER_ACTIVE_VALIDATORS:
+        raise Exception(f"Invalid mainchainValidators property. It must have exactly {MAINCHAIN_NUMBER_ACTIVE_VALIDATORS} elements.")
+    
+    validatorKeys = [validator["blsKey"] for validator in trs.params.mainchainValidators]
+    # All validator keys must be distinct.
+    if len(validatorKeys) != len(set(validatorKeys)):
+        raise Exception("Duplicate BLS keys.")
+    
+    # Validator keys must be in lexicographic order.
+    if not all(validatorKeys[i] < validatorKeys[i + 1] for i in range(len(validatorKeys) - 1)):
+        raise Exception("Validator keys are not in lexicographic order.")
+    
+    for validator in trs.params.initValidators:
+        # The bftWeight property of each element is equal to 1.
+        if validator["bftWeight"] != 1:
+            raise Exception("Invalid bftWeight property.")
+```
+
+
+#### Execution
+
+```python
+def execute(trs: Transaction) -> None:
+    # Check signature property.
+    height = height of the block in which trs was included
+    validators = BFT.getBFTParameters(b.header.height).validators
+    certificateThreshold = BFT.getBFTParameters(b.header.height).certificateThreshold
+    # Sort validators by BLS key.
+    validators.sort(key=lambda validator: validatorsModule.getValidatorAccount(validator.address).blsKey)
```

Comment:
The validators are guaranteed to be sorted by the key in the `verify()` function, I think.

Reply:
in the verify() step we check that the mainchain validators given in the trs.params are in lexicographical order, here instead we get the sidechain validators from the BFT module